File: snippets/julia/CumSum.jl

function cumsum_1d(arr::Vector{T})::Vector{T} where {T}
    s = [0]
    for i in 0:length(arr)
        push!(s, s[i] + arr[i])
    end
    return s
end

function sum(cumsum::Vector, s::Int, e::Int)::Int
    return cumsum[e] - cumsum[s]
end

function cumsum_2d(mat::Matrix{T})::Matrix where {T}
    s = zeros(size(mat)[1] + 1, size(mat)[2] + 1)
    for y in 1:size(mat)[1]
        for x in 1:size(mat)[2]
            s[x + 1, y + 1] = s[x, y + 1] + s[x + 1, y] - s[x, y] + mat[x, y]
        end
    end
    return s
end

function sum(cumsum::Matrix{T}, s::Tuple{Int, Int}, e::Tuple{Int, Int})::Int where {T}
    return cumsum[e[1], e[2]] - cumsum[s[1], e[2]] - cumsum[e[1], s[2]] + cumsum[s[1], s[2]]
end


function test_cumsum()
    cs1 = cumsum_1d([1, 3, 4, 2, 5])
    @assert sum(cs1, 2, 4) == 6
    @assert sum(cs1, 0, 5) == 15

    cs2 = cumsum_2d([
        1 8 7 3 2
        9 1 3 4 6
        3 5 8 1 4
        2 7 3 2 5
    ])
    @assert sum(cs2, (1, 2), (3, 5)) == 26
    @assert sum(cs2, (0, 1), (2, 3)) == 19
    @assert sum(cs2, (0, 0), (4, 5)) == 84
end

test_cumsum()

File: snippets/python/cumsum.py

# 1D cumsum
class CumSum1D:
    def __init__(self, arr: list):
        self.s = [0]
        l = len(arr)
        for i in range(l):
            self.s.append(self.s[i] + arr[i])

    def sum(self, s: int, e: int) -> int:
        """ [s, e) """
        return self.s[e] - self.s[s]

# 2D cumsum
class CumSum2D:
    def __init__(self, grid: list) -> None:
        w, h = len(grid[0]), len(grid)
        self.s = [[0 for i in range(w + 1)] for j in range(h + 1)]
        for i in range(h):
            for j in range(w):
                self.s[i + 1][j + 1] = self.s[i][j + 1] + self.s[i + 1][j] - self.s[i][j] + grid[i][j]

    def sum(self, s: tuple, e: tuple) -> int:
        """ [sx, ex) * [sy, ey) """
        return self.s[e[0]][e[1]] - self.s[s[0]][e[1]] - self.s[e[0]][s[1]] + self.s[s[0]][s[1]]

def test_cumsum():
    cs1 = CumSum1D([1, 3, 4, 2, 5])
    assert(cs1.sum(2, 4) == 6)
    assert(cs1.sum(0, 5) == 15)

    cs2 = CumSum2D([
        [1, 8, 7, 3, 2],
        [9, 1, 3, 4, 6],
        [3, 5, 8, 1, 4],
        [2, 7, 3, 2, 5]
    ])
    assert(cs2.sum((1, 2), (3, 5)) == 26)
    assert(cs2.sum((0, 1), (2, 3)) == 19)
    assert(cs2.sum((0, 0), (4, 5)) == 84)

test_cumsum()
